Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp

```diff
@@ -0,0 +1,238 @@
+#include "RenderElement.h"
+#include "RenderStyleInlines.h"
+#include "RenderTreeUpdater.h"
+#include "RenderView.h"
+#include "StyleTreeResolver.h"
+#include "ViewTransition.h"
+
+namespace WebCore {
+
+RenderTreeUpdater::ViewTransition::ViewTransition(RenderTreeUpdater& updater)
+    : m_updater(updater)
+{
+}
+
+// The contents and ordering of the named elements map should remain stable during the duration of the transition.
+// We should only need to handle changes in the `display` CSS property by recreating / deleting renderers as needed.
+void RenderTreeUpdater::ViewTransition::updatePseudoElementTree(RenderElement& documentElementRenderer)
+{
+    auto destroyPseudoElementTreeIfNeeded = [&documentElementRenderer, this]() {
+        if (WeakPtr viewTransitionRoot = documentElementRenderer.view().viewTransitionRoot())
+            m_updater.m_builder.destroy(*viewTransitionRoot);
+    };
+
+    Ref document = documentElementRenderer.document();
+
+    // Intentionally bail out early here to avoid computing the style.
+    if (!document->hasViewTransitionPseudoElementTree() || !document->documentElement()) {
+        destroyPseudoElementTreeIfNeeded();
+        return;
+    }
+
+    // Destroy pseudo element tree ::view-transition has display: none or no style.
+    auto rootStyle = documentElementRenderer.getCachedPseudoStyle(PseudoId::ViewTransition, &documentElementRenderer.style());
+    if (!rootStyle || rootStyle->display() == DisplayType::None) {
+        destroyPseudoElementTreeIfNeeded();
+        return;
+    }
+
+    RefPtr activeViewTransition = document->activeViewTransition();
+    ASSERT(activeViewTransition);
+
+    auto newRootStyle = RenderStyle::clone(*rootStyle);
+
+    // Create ::view-transition as needed.
+    SingleThreadWeakPtr<RenderElement> viewTransitionRoot;
+    if ((viewTransitionRoot = documentElementRenderer.view().viewTransitionRoot().get()))
+        viewTransitionRoot->setStyle(WTFMove(newRootStyle));
+    else {
+        auto newViewTransitionRoot = WebCore::createRenderer<RenderBlockFlow>(RenderObject::Type::BlockFlow, documentElementRenderer.document(), WTFMove(newRootStyle));
+        newViewTransitionRoot->initializeStyle();
+        documentElementRenderer.view().setViewTransitionRoot(*newViewTransitionRoot.get());
+        viewTransitionRoot = newViewTransitionRoot.get();
+        m_updater.m_builder.attach(documentElementRenderer.view(), WTFMove(newViewTransitionRoot));
+    }
+
+    // No groups. The map is constant during the duration of the transition, so we don't need to handle deletions.
+    if (activeViewTransition->namedElements().isEmpty())
+        return;
+
+    // Traverse named elements map to update/build all ::view-transition-group().
+    Vector<SingleThreadWeakPtr<RenderObject>> descendantsToDelete;
+    auto* currentGroup = documentElementRenderer.view().viewTransitionRoot()->firstChild();
+    for (auto& name : activeViewTransition->namedElements().keys()) {
```

Comment:
`const auto&`?

Reply:
I don't think it makes a huge difference, but I could use it